File: cli/internal/connectapi/threads.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/types/known/timestamppb"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
)

const (
	defaultFollowedThreadLimit = 20
	maxFollowedThreadLimit     = 100
)

type threadService struct {
	api *API
}

func (s *threadService) ListFollowedThreads(ctx context.Context, req *connect.Request[apiv1.ListFollowedThreadsRequest]) (*connect.Response[apiv1.ListFollowedThreadsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	limit, offset := apiPagination(req.Msg.GetPage(), defaultFollowedThreadLimit, maxFollowedThreadLimit)
	page, err := s.api.core.ThreadFollows().ListFollowedThreads(ctx, caller.UserID, limit, offset)
	if err != nil {
		return nil, connectError(err)
	}

	resp, err := s.followedThreadsResponse(ctx, caller.UserID, page)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(resp), nil
}

func (s *threadService) FollowThread(ctx context.Context, req *connect.Request[apiv1.FollowThreadRequest]) (*connect.Response[apiv1.FollowThreadResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	if err := s.api.core.ThreadFollows().FollowThread(ctx, caller.UserID, req.Msg.RoomId, req.Msg.ThreadRootEventId); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.FollowThreadResponse{Following: true}), nil
}

func (s *threadService) UnfollowThread(ctx context.Context, req *connect.Request[apiv1.UnfollowThreadRequest]) (*connect.Response[apiv1.UnfollowThreadResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	if err := s.api.core.ThreadFollows().UnfollowThread(ctx, caller.UserID, req.Msg.RoomId, req.Msg.ThreadRootEventId); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.UnfollowThreadResponse{Following: false}), nil
}

func (s *threadService) followedThreadsResponse(ctx context.Context, viewerID string, page *core.FollowedThreadsPage) (*apiv1.ListFollowedThreadsResponse, error) {
	if page == nil {
		return &apiv1.ListFollowedThreadsResponse{
			Includes: &apiv1.RoomTimelineIncludes{Users: map[string]*apiv1.User{}},
			Page:     apiPageInfo(0, false),
		}, nil
	}

	messageIDs := make([]string, 0, len(page.Threads))
	for _, thread := range page.Threads {
		if thread != nil && thread.ThreadRootEventID != "" {
			messageIDs = append(messageIDs, thread.ThreadRootEventID)
		}
	}
	reactionsByMessageID, err := s.api.core.GetReactionsBatch(ctx, messageIDs)
	if err != nil {
		return nil, err
	}

	h := &timelineHydrator{
		api:                  s.api,
		ctx:                  ctx,
		viewerID:             viewerID,
		kind:                 core.KindChannel,
		reactionsByMessageID: reactionsByMessageID,
		userIDs:              make(map[string]struct{}),
	}

	threads := make([]*apiv1.FollowedThread, 0, len(page.Threads))
	for _, thread := range page.Threads {
		if thread == nil {
			continue
		}

		kind := core.RoomKindFromLegacySpaceID(thread.SpaceID)
		room, err := s.api.core.GetRoom(ctx, kind, thread.RoomID)
		if err != nil {
			return nil, err
		}

		event, err := s.api.core.GetRoomEventByEventID(ctx, kind, thread.RoomID, thread.ThreadRootEventID)
		if err != nil {
			return nil, err
		}
		var rootMessage *apiv1.RoomTimelineEvent
		if event != nil {
			rootMessage, err = h.event(&core.RoomEvent{Event: event})
			if err != nil {
				return nil, err
			}
		}

		var lastReplyAt *timestamppb.Timestamp
		if thread.LastReplyAt != nil {
			lastReplyAt = timestamppb.New(*thread.LastReplyAt)
		}
		threads = append(threads, &apiv1.FollowedThread{
			RoomId:            thread.RoomID,
			RoomName:          room.GetName(),
			ThreadRootEventId: thread.ThreadRootEventID,
			RootMessage:       rootMessage,
			ReplyCount:        int32(thread.ReplyCount),
			LastReplyAt:       lastReplyAt,
			HasUnread:         thread.HasUnread,
		})
	}

	users, err := h.users()
	if err != nil {
		return nil, err
	}

	return &apiv1.ListFollowedThreadsResponse{
		Threads:  threads,
		Includes: &apiv1.RoomTimelineIncludes{Users: users},
		Page:     apiPageInfo(page.TotalCount, page.HasMore),
	}, nil
}

File: apps/frontend/src/lib/api-client/threads.ts
import { authHeaders, createChattoClient, handleAuthError } from "./connect.js";
import { ThreadService } from "@chatto/api-types/api/v1/threads_connect";
import type { User } from "@chatto/api-types/api/v1/users_pb";
import type { RawEvent } from "./events.js";
import { roomTimelineEventToRawEvent } from "./roomTimeline.js";

export type ConnectAPIConfig = {
  serverId?: string;
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type FollowedThread = {
  roomId: string;
  roomName: string;
  threadRootEventId: string;
  rootMessage: RawEvent | null;
  replyCount: number;
  lastReplyAt: string | null;
  hasUnread: boolean;
};

export type FollowedThreadsPage = {
  threads: FollowedThread[];
  totalCount: number;
  hasMore: boolean;
};

export type ThreadFollowState = {
  roomId: string;
  threadRootEventId: string;
  following: boolean;
};

export type ThreadFollowResult = {
  following: boolean;
  state: ThreadFollowState | null;
};

export function createThreadAPI(config: ConnectAPIConfig) {
  const client = createChattoClient(ThreadService, config);
  const headers = () => authHeaders(config);
  return {
    async listFollowedThreads(input: {
      limit: number;
      offset: number;
    }): Promise<FollowedThreadsPage> {
      try {
        const response = await client.listFollowedThreads(
          { page: { limit: input.limit, offset: input.offset } },
          { headers: headers() },
        );
        const users = response.includes?.users ?? {};
        return {
          threads: response.threads.map((thread) => ({
            roomId: thread.roomId,
            roomName: thread.roomName,
            threadRootEventId: thread.threadRootEventId,
            rootMessage: thread.rootMessage
              ? roomTimelineEventToRawEvent(
                  thread.rootMessage,
                  users as Record<string, User>,
                )
              : null,
            replyCount: thread.replyCount,
            lastReplyAt: timestampToISOOrNull(thread.lastReplyAt),
            hasUnread: thread.hasUnread,
          })),
          totalCount: Number(response.page?.totalCount ?? 0),
          hasMore: response.page?.hasMore ?? false,
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async followThread(input: {
      roomId: string;
      threadRootEventId: string;
    }): Promise<ThreadFollowResult> {
      try {
        const response = await client.followThread(input, {
          headers: headers(),
        });
        return {
          following: response.following,
          state: response.state ? mapThreadFollowState(response.state) : null,
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async unfollowThread(input: {
      roomId: string;
      threadRootEventId: string;
    }): Promise<ThreadFollowResult> {
      try {
        const response = await client.unfollowThread(input, {
          headers: headers(),
        });
        return {
          following: response.following,
          state: response.state ? mapThreadFollowState(response.state) : null,
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },
  };
}

function mapThreadFollowState(state: {
  roomId: string;
  threadRootEventId: string;
  following: boolean;
}): ThreadFollowState {
  return {
    roomId: state.roomId,
    threadRootEventId: state.threadRootEventId,
    following: state.following,
  };
}

function timestampToISOOrNull(
  timestamp: { toDate(): Date } | undefined,
): string | null {
  return timestamp ? timestamp.toDate().toISOString() : null;
}
